### PR TITLE
is_set_to: unify TH/THC implmentation and genericize test_is_set_to.

### PR DESCRIPTION
**Pull request by gchanan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28514 Kill zero_dim_tensor_only codegen, it's not used anymore.
* #28513 Remove zero_dim_tensors only from _th_masked_fill_.
* #28512 Remove zero_dim_tensors only from _th_index_fill_.
* #28511 Kill _th_fill binding, which isn't used anymore.
* #28510 unfold: turn off device_guard
* #28475 Port TH/THC implementation of unfold to ATen.
* #28462 Fix scalar handling of unfold.
* #28435 Kill TH(C)Tensor_squeeze which isn't used anymore.
* #28425 Port is_set_to from TH/THC to ATen.
* **#28422 is_set_to: unify TH/THC implmentation and genericize test_is_set_to.**

The TH implementation had two differences:
1) It explicitly checked for null storages; this isn't supported anymore so can be removed.
2) It collapsed all empty tensors to the same shape for the purpose of checking.  This was introduced to keep BC when we introduced N-dimensional empty tensors,
but since it's been quite a long time since we've had N-dimensional empty tensors and the CUDA implementation didn't support this, we should get rid of it.

Differential Revision: [D18061916](https://our.internmc.facebook.com/intern/diff/D18061916)

### CODE REVIEW

File: aten/src/TH/generic/THTensor.cpp [hunk 1]
```diff
@@ -527,14 +527,12 @@ int THTensor_(isSameSizeAs)(const THTensor *self, const THTensor* src)
...
-  if (!THTensor_getStoragePtr(self))
-    return 0;
   if (THTensor_getStoragePtr(self) == THTensor_getStoragePtr(src) &&
       self->storage_offset() == src->storage_offset() &&
-      THTensor_nDimensionLegacyAll(self) == THTensor_nDimensionLegacyAll(src))
+      THTensor_nDimension(self) == THTensor_nDimension(src))
   {
     int d;
-    for (d = 0; d < THTensor_nDimensionLegacyAll(self); ++d)
+    for (d = 0; d < THTensor_nDimension(self); ++d)
```
Comment (mruberry) [inline aten/src/TH/generic/THTensor.cpp]: auto d = decltype(THTensor_nDimension(self)){0}

File: test/test_torch.py [hunk 2]
```diff
@@ -6370,6 +6353,29 @@ def test_cat_empty(self, device):
         self.assertRaises(RuntimeError, lambda: torch.cat([x, empty], dim=1))
         self.assertRaises(RuntimeError, lambda: torch.cat([empty, x], dim=1))
 
+    def test_is_set_to(self, device):
```
Comment (mruberry) [inline test/test_torch.py]: This test case may need to be updated per changes to the prior PR.

**Review by mruberry — APPROVED** [review]
Looks good. May need minor edits based on changes to stacked PR.

File: aten/src/TH/generic/THTensor.cpp [hunk 3]
```diff
@@ -527,14 +527,12 @@ int THTensor_(isSameSizeAs)(const THTensor *self, const THTensor* src)
 
 int THTensor_(isSetTo)(const THTensor *self, const THTensor* src)
 {
-  if (!THTensor_getStoragePtr(self))
```
Comment (albanD) [inline aten/src/TH/generic/THTensor.cpp]: Won't the next `if` segfault if both storages are nullptr? 
Or are the storage supposed to never be nullptr anymore?

Comment (gchanan) [reply aten/src/TH/generic/THTensor.cpp]: They aren't supposed to be nullptr anymore (in TH).

**Pull request closed without merging**